lib/oplog/tail.go

```diff
@@ -529,3 +544,10 @@ func parseNamespace(namespace string) (string, string) {
 
 	return database, collection
 }
+
+// assignToShard determines which shard should process a given key.
+// This should just be key % shardCount, but Go modulo is weird with negative numbers,
+// and the parallelism key can be negative.
+func assignToShard(key int, shardCount int) int {
+	return (key%shardCount + shardCount) % shardCount
+}
```

Comment:
It's not clear to me from the comment where the negative numbers manifest and how this solves it

Reply:
I'll update the comment:

essentially, we want to take the "key" which is random bytes, and get an index into the shard array (length=N). ideally we would just do key%N, since in a sane world that would map to the range [0, N) which is the set of array indices. However because the key is random bytes, it can be negative, and Go is weird where (e.g.) `-3%5` is `-3` instead of `2`. so you get array index out of bounds. we fix this by doing `(a%b+b)%b` instead of `a%b` which works as expected.